src/reflect/scala/reflect/internal/Depth.scala

```diff
@@ -37,4 +39,9 @@ object Depth {
     if (depth < AnyDepthValue) AnyDepth
     else new Depth(depth)
   }
+
+  def maximumBy[A](xs: List[A])(ff: DepthFunction[A]): Depth = {
+    def maxDepth(d: Depth, x: A) = d max ff(x)
+    xs.foldLeft(Zero)(maxDepth)
+  }
```

Comment:
This needs to box the `Depth` value classes to work with the generic code of `List.foldLeft`. Use a `while` loop instead.

```
  public maximumBy(Lscala/collection/immutable/List;Lscala/reflect/internal/DepthFunction;)I
    // parameter final  xs
    // parameter final  ff
    ALOAD 1
    NEW scala/reflect/internal/Depth
    DUP
    ALOAD 0
    INVOKEVIRTUAL scala/reflect/internal/Depth$.Zero ()I
    INVOKESPECIAL scala/reflect/internal/Depth.<init> (I)V
    ALOAD 2
    INVOKEDYNAMIC apply(Lscala/reflect/internal/DepthFunction;)Lscala/Function2; [
      // handle kind 0x6 : INVOKESTATIC
      java/lang/invoke/LambdaMetafactory.altMetafactory(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/MethodType;[Ljava/lang/Object;)Ljava/lang/invoke/CallSite;
      // arguments:
      (Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;,
      // handle kind 0x6 : INVOKESTATIC
      scala/reflect/internal/Depth$.$anonfun$maximumBy$1$adapted(Lscala/reflect/internal/DepthFunction;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;,
      (Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;,
      3,
      1,
      java.io.Serializable.class
    ]
    ASTORE 4
    ASTORE 3
    IFNONNULL L0
    ACONST_NULL
    ATHROW
   L0
    ALOAD 1
    ALOAD 3
    ALOAD 4
    INVOKESTATIC scala/collection/LinearSeqOps.foldLeft$ (Lscala/collection/LinearSeqOps;Ljava/lang/Object;Lscala/Function2;)Ljava/lang/Object; (itf)
    CHECKCAST scala/reflect/internal/Depth
    INVOKEVIRTUAL scala/reflect/internal/Depth.depth ()I
    IRETURN
    MAXSTACK = 4
    MAXLOCALS = 5

  // access flags 0x1019
  public final static synthetic $anonfun$maximumBy$1(Lscala/reflect/internal/DepthFunction;ILjava/lang/Object;)I
    // parameter final  ff$1
    // parameter final  d
    // parameter final  x
    ILOAD 1
    ALOAD 2
    ALOAD 0
    INVOKESTATIC scala/reflect/internal/Depth$.maxDepth$1 (ILjava/lang/Object;Lscala/reflect/internal/DepthFunction;)I
    IRETURN
    MAXSTACK = 3
    MAXLOCALS = 3

  // access flags 0x1019
  public final static synthetic $anonfun$maximumBy$1$adapted(Lscala/reflect/internal/DepthFunction;Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
    // parameter final  ff$1
    // parameter final  d
    // parameter final  x
    NEW scala/reflect/internal/Depth
    DUP
    ALOAD 0
    ALOAD 1
    CHECKCAST scala/reflect/internal/Depth
    INVOKEVIRTUAL scala/reflect/internal/Depth.depth ()I
    ALOAD 2
    INVOKESTATIC scala/reflect/internal/Depth$.$anonfun$maximumBy$1 (Lscala/reflect/internal/DepthFunction;ILjava/lang/Object;)I
    INVOKESPECIAL scala/reflect/internal/Depth.<init> (I)V
    ARETURN
    MAXSTACK = 5
    MAXLOCALS = 3
```